Source/Core/VideoCommon/ShaderCache.cpp

```diff
@@ -157,32 +157,43 @@ const AbstractPipeline* ShaderCache::GetUberPipelineForUid(const GXUberPipelineU
 
 void ShaderCache::WaitForAsyncCompiler()
 {
-  while (m_async_shader_compiler->HasPendingWork() || m_async_shader_compiler->HasCompletedWork())
-  {
-    m_async_shader_compiler->WaitUntilCompletion([](size_t completed, size_t total) {
-      g_renderer->BeginUIFrame();
-
-      const float center_x = ImGui::GetIO().DisplaySize.x * 0.5f;
-      const float center_y = ImGui::GetIO().DisplaySize.y * 0.5f;
-      const float scale = ImGui::GetIO().DisplayFramebufferScale.x;
-
-      ImGui::SetNextWindowSize(ImVec2(400.0f * scale, 50.0f * scale), ImGuiCond_Always);
-      ImGui::SetNextWindowPos(ImVec2(center_x, center_y), ImGuiCond_Always, ImVec2(0.5f, 0.5f));
-      if (ImGui::Begin(Common::GetStringT("Compiling Shaders").c_str(), nullptr,
-                       ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoInputs |
-                           ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoSavedSettings |
-                           ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoNav |
-                           ImGuiWindowFlags_AlwaysAutoResize | ImGuiWindowFlags_NoFocusOnAppearing))
-      {
-        ImGui::Text("Compiling shaders: %zu/%zu", completed, total);
-        ImGui::ProgressBar(static_cast<float>(completed) /
-                               static_cast<float>(std::max(total, static_cast<size_t>(1))),
-                           ImVec2(-1.0f, 0.0f), "");
-      }
-      ImGui::End();
+  bool running = true;
 
-      g_renderer->EndUIFrame();
-    });
+  while (running &&
```

Comment:
I propose this:

```
constexpr auto update_ui_progress = [](size_t completed, size_t total) {
	g_renderer->BeginUIFrame();

	const float center_x = ImGui::GetIO().DisplaySize.x * 0.5f;
	const float center_y = ImGui::GetIO().DisplaySize.y * 0.5f;
	const float scale = ImGui::GetIO().DisplayFramebufferScale.x;

	ImGui::SetNextWindowSize(ImVec2(400.0f * scale, 50.0f * scale), ImGuiCond_Always);
	ImGui::SetNextWindowPos(ImVec2(center_x, center_y), ImGuiCond_Always, ImVec2(0.5f, 0.5f));
	if (ImGui::Begin(Common::GetStringT("Compiling Shaders").c_str(), nullptr,
				ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoInputs |
					ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoSavedSettings |
					ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoNav |
					ImGuiWindowFlags_AlwaysAutoResize |
					ImGuiWindowFlags_NoFocusOnAppearing))
	{
		ImGui::Text("Compiling shaders: %zu/%zu", completed, total);
		ImGui::ProgressBar(static_cast<float>(completed) /
						static_cast<float>(std::max(total, static_cast<size_t>(1))),
					ImVec2(-1.0f, 0.0f), "");
	}
	ImGui::End();

	g_renderer->EndUIFrame();
}

while(m_async_shader_compiler->HasPendingWork() || m_async_shader_compiler->HasCompletedWork())
{
	if (!m_async_shader_compiler->WaitUntilCompletion(update_ui_progress))
		break;

	m_async_shader_compiler->RetrieveWorkItems();
}

// Just render nothing to clear the screen
g_renderer->BeginUIFrame();
g_renderer->EndUIFrame();
```

My reasoning but maybe I missed something:

1)  Breaking out the ui update into a lambda makes it a little readable imo.
2)  If you are cancelling, I don't think you need to wait on the other work items
3)  Passing the lambda for the clear is not necessary, just call that after the function returns.  Additionally, I don't think you need to call it every iteration, just the last one.

Reply:
(also included @Dentomologist 's suggestion about removing `interruptable` )

Reply:
`StopWorkerThreads()` [joins on all worker threads](https://github.com/dolphin-emu/dolphin/blob/f4e2d369c075f44f037289a1ccb78274c7d5cb39/Source/Core/VideoCommon/AsyncShaderCompiler.cpp#L174-L177), thus waits for them to gracefully quit. This effectively means we wait for the currently started chunks of work to finish, then discard the remainder of that pending work.

I'll address the other points, I like the idea of this lambda being separate indeed.